feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/ui/RotationUtil.kt

```diff
@@ -0,0 +1,118 @@
+import androidx.core.view.WindowInsetsControllerCompat
+
+/**
+ * As long as this composable is active, the window will go into immersive mode and prevents the
+ * rotation animation on configuration change. This will prevent the UI items from visually changing.
+ *
+ * When used in combination with a composable that renders the same UI in both landscape and portrait,
+ * it can create a smooth continuous feel between those two orientations.
+ */
+@Composable
+fun SmoothImmersiveRotationEffect(context: Context) = DisposableEffect(context) {
+    var currentRotationAnimation: Int? = null
+    context.getActivity()?.window?.let { window ->
+        window.attributes = window.attributes.apply {
+            currentRotationAnimation = rotationAnimation
+            rotationAnimation = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                ROTATION_ANIMATION_SEAMLESS
+            } else {
+                ROTATION_ANIMATION_JUMPCUT
+            }
+        }
+        WindowCompat.getInsetsController(window, window.decorView).apply {
+            systemBarsBehavior =
+                WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
+            hide(WindowInsetsCompat.Type.systemBars())
+        }
+    }
+    onDispose {
+        context.getActivity()?.window?.let { window ->
+            if (currentRotationAnimation != null) {
+                window.attributes = window.attributes.apply {
+                    rotationAnimation = WindowManager.LayoutParams.ROTATION_ANIMATION_ROTATE
+                }
+            }
+            WindowCompat.getInsetsController(window, window.decorView).apply {
+                show(WindowInsetsCompat.Type.systemBars())
+            }
+        }
+    }
+}
+
+/**
+ * Rotate a layout based on the current screen orientation. The UI will always be laid out in a way
+ * that width <= height, and rotated afterwards.
+ */
+@Composable
+fun Modifier.rotatedLayout(): Modifier {
+    var currentOrientation by remember { mutableIntStateOf(Surface.ROTATION_0) }
+    val currentDegrees = currentOrientation * 90f
+    val newOrientation = LocalConfiguration.current.orientation
+    val display = LocalView.current.display
+    LaunchedEffect(newOrientation, display) {
+        val newRotation = display.rotation
+        if (currentOrientation != newRotation) {
+            currentOrientation = newRotation
+        }
+    }
+    return this then Modifier.fillMaxSize().layout { measurable, constraints ->
+        val height = maxOf(constraints.maxWidth, constraints.maxHeight)
+        val width = minOf(constraints.maxWidth, constraints.maxHeight)
+        val placeable = measurable.measure(
+            Constraints.fixed(width, height)
+        )
+        layout(placeable.width, placeable.height) {
+            placeable.placeWithLayer(0, 0) {
+                if (constraints.maxWidth > constraints.maxHeight) {
+                    rotationZ = -currentDegrees
```

Comment:
I think this logic assumes that a `Display.rotation` of `0` implies a portrait orientation. On the Pixel Fold inner screen, `Display.rotation` is `0` when in landscape, so this will layout the content with width <= height, but then not rotate it by any amount.

Reply:
Good catch. I've changed the code to base the rotation on what the layout should look like in the device natural orientation.